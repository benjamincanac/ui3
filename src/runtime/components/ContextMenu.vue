<script lang="ts">
import { tv } from 'tailwind-variants'
import type { ContextMenuRootProps, ContextMenuRootEmits, ContextMenuContentProps, ContextMenuArrowProps } from 'radix-vue'
import type { AppConfig } from '@nuxt/schema'
import _appConfig from '#build/app.config'
import theme from '#build/ui/context-menu'
import type { AvatarProps, KbdProps, LinkProps } from '#ui/types'
import type { DynamicSlots } from '#ui/types/utils'

const appConfig = _appConfig as AppConfig & { ui: { contextMenu: Partial<typeof theme> } }

const contextMenu = tv({ extend: tv(theme), ...(appConfig.ui?.contextMenu || {}) })

export interface ContextMenuItem extends Omit<LinkProps, 'type'> {
  label?: string
  icon?: string
  avatar?: AvatarProps
  content?: Omit<ContextMenuContentProps, 'asChild' | 'forceMount'>
  kbds?: KbdProps['value'][] | KbdProps[]
  /**
   * The item type.
   * @defaultValue 'link'
   */
  type?: 'label' | 'separator' | 'link'
  slot?: string
  open?: boolean
  defaultOpen?: boolean
  children?: ContextMenuItem[] | ContextMenuItem[][]
  select? (e: Event): void
}

export interface ContextMenuProps<T> extends Omit<ContextMenuRootProps, 'dir'> {
  items?: T[] | T[][]
  disabled?: boolean
  content?: Omit<ContextMenuContentProps, 'asChild' | 'forceMount'>
  arrow?: boolean | Omit<ContextMenuArrowProps, 'asChild'>
  portal?: boolean
  class?: any
  ui?: Partial<typeof contextMenu.slots>
}

export interface ContextMenuEmits extends ContextMenuRootEmits {}

type SlotProps<T> = (props: { item: T, active?: boolean, index: number }) => any

export type ContextMenuSlots<T extends { slot?: string }> = {
  default(): any
  leading: SlotProps<T>
  label: SlotProps<T>
  trailing: SlotProps<T>
  item: SlotProps<T>
} & DynamicSlots<T, SlotProps<T>>
</script>

<script setup lang="ts" generic="T extends ContextMenuItem">
import { computed, toRef } from 'vue'
import { ContextMenuRoot, ContextMenuTrigger, ContextMenuArrow, useForwardPropsEmits } from 'radix-vue'
import { reactivePick } from '@vueuse/core'
import { UContextMenuContent } from '#components'
import { omit } from '#ui/utils'

const props = withDefaults(defineProps<ContextMenuProps<T>>(), {
  portal: true,
  modal: false
})
const emits = defineEmits<ContextMenuEmits>()
const slots = defineSlots<ContextMenuSlots<T>>()

const rootProps = useForwardPropsEmits(reactivePick(props, 'modal'), emits)
const contentProps = toRef(() => props.content as ContextMenuContentProps)
const arrowProps = toRef(() => props.arrow as ContextMenuArrowProps)
const proxySlots = omit(slots, ['default']) as Record<string, ContextMenuSlots<T>[string]>

const ui = computed(() => tv({ extend: contextMenu, slots: props.ui })())
</script>

<template>
  <ContextMenuRoot v-bind="rootProps">
    <ContextMenuTrigger v-if="$slots.default" as-child :disabled="disabled">
      <slot />
    </ContextMenuTrigger>

    <UContextMenuContent :class="ui.content({ class: props.class })" :ui="ui" v-bind="contentProps" :items="items" :portal="portal">
      <template v-for="(_, name) in proxySlots" #[name]="slotData: any">
        <slot :name="name" v-bind="slotData" />
      </template>

      <ContextMenuArrow v-if="!!arrow" v-bind="arrowProps" :class="ui.arrow()" />
    </UContextMenuContent>
  </ContextMenuRoot>
</template>
