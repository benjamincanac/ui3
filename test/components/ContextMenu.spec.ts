import { describe, it, expect } from 'vitest'
import ContextMenu, { type ContextMenuProps, type ContextMenuSlots } from '../../src/runtime/components/ContextMenu.vue'
import ComponentRender from '../component-render'

// FIXME: Can't force open state
describe('ContextMenu', () => {
  const items = [
    [{
      label: 'Appearance',
      children: [{
        label: 'System',
        icon: 'i-heroicons-computer-desktop'
      }, {
        label: 'Light',
        icon: 'i-heroicons-sun'
      }, {
        label: 'Dark',
        icon: 'i-heroicons-moon'
      }]
    }],
    [{
      label: 'Show Sidebar',
      kbds: ['meta', 'S'],
      select() {
        console.log('Show Sidebar clicked')
      }
    }, {
      label: 'Show Toolbar',
      kbds: ['shift', 'meta', 'D'],
      select() {
        console.log('Show Toolbar clicked')
      }
    }, {
      label: 'Collapse Pinned Tabs',
      disabled: true
    }], [{
      label: 'Refresh the Page'
    }, {
      label: 'Clear Cookies and Refresh'
    }, {
      label: 'Clear Cache and Refresh'
    }, {
      type: 'separator' as const
    }, {
      label: 'Developer',
      children: [[{
        label: 'View Source',
        kbds: ['option', 'meta', 'U'],
        select() {
          console.log('View Source clicked')
        }
      }, {
        label: 'Developer Tools',
        kbds: ['option', 'meta', 'I'],
        select() {
          console.log('Developer Tools clicked')
        }
      }], [{
        label: 'Inspect Elements',
        kbds: ['option', 'meta', 'C'],
        select() {
          console.log('Inspect Elements clicked')
        }
      }], [{
        label: 'JavaScript Console',
        kbds: ['option', 'meta', 'J'],
        slot: 'custom',
        select() {
          console.log('JavaScript Console clicked')
        }
      }]]
    }]
  ]

  const props = { portal: false, items }

  it.each([
    // Props
    ['with items', { props }],
    ['with disabled', { props: { ...props, disabled: true } }],
    ['with class', { props: { ...props, class: 'min-w-96' } }],
    ['with ui', { props: { ...props, ui: { linkLeadingIcon: 'size-4' } } }],
    // Slots
    ['with default slot', { props, slots: { default: () => 'Default slot' } }],
    ['with leading slot', { props, slots: { leading: () => 'Leading slot' } }],
    ['with label slot', { props, slots: { label: () => 'Label slot' } }],
    ['with trailing slot', { props, slots: { trailing: () => 'Trailing slot' } }],
    ['with item slot', { props, slots: { item: () => 'Item slot' } }],
    ['with custom slot', { props, slots: { custom: () => 'Custom slot' } }]
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: ContextMenuProps<typeof items[number][number]>, slots?: Partial<ContextMenuSlots<any>> }) => {
    const html = await ComponentRender(nameOrHtml, options, ContextMenu)
    expect(html).toMatchSnapshot()
  })
})
