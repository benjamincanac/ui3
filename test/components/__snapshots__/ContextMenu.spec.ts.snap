// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContextMenu > renders with class correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with custom slot correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with default slot correctly 1`] = `
"Default slot
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with disabled correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with item slot correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with items correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with label slot correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with leading slot correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with trailing slot correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;

exports[`ContextMenu > renders with ui correctly 1`] = `
"<!--v-if-->
<!--teleport start-->



<!---->



<!--teleport end-->"
`;
